### PR TITLE
acrn-dm cpu affinity dynamic param refine

### PR DESCRIPTION
**Pull request by Kunhui-Li**
The GET_PLATFORM_INFO will be removed from hypervisor. The acrn-dm can only refer to Service VM's resources from now, all the resources out of Service VM are not awared by acrn-dm. The original info got from GET_PLATFORM_INFO needs to changed to Service VM's perspective like lapic id.

dm: 
1. parse the Service VM's MADT table to convert the pcpu_id to lapic_id instead of GET_PLATFORM_INFO.
2. use lapic id to set cpu affinity.

config_tools:
1. use apic_id got from board XML file instead of pcpu id from scenario XML file according to the refined usage '--cpu_affinity lapic_id'.

### CODE REVIEW

**Comment by acrnsi** [comment]
Can one of the admins verify this patch?

**Review by NanlinXie — APPROVED** [review]
Approve for configtool part.

File: devicemodel/core/main.c [hunk 1]
```diff
@@ -166,7 +166,8 @@ usage(int code)
 		"       --mac_seed: set a platform unique string as a seed for generate mac address\n"
 		"       --ovmf: ovmf file path\n"
 		"       --ssram: Congfiure Software SRAM parameters\n"
-		"       --cpu_affinity: list of pCPUs assigned to this VM\n"
+		"       --cpu_affinity: list of Service VM vcpus assigned to this VM, the Service VM vcpus are"
```
Comment (gvancuts) [inline devicemodel/core/main.c:169]:
```suggestion
		"       --cpu_affinity: list of Service VM vCPUs assigned to this User VM, the vCPUs are"
```

File: devicemodel/core/main.c [hunk 2]
```diff
@@ -166,7 +166,8 @@ usage(int code)
...
+		"	     identified by their lapic ids.\n"
```
Comment (gvancuts) [inline devicemodel/core/main.c:170]:
```suggestion
		"	     identified by their local APIC IDs.\n"
```

File: devicemodel/core/main.c [hunk 3]
```diff
@@ -845,6 +846,11 @@ main(int argc, char *argv[])
 	if (signal(SIGPIPE, SIG_IGN) == SIG_ERR)
 		fprintf(stderr, "cannot register handler for SIGPIPE\n");
 
+	if (parse_madt()) {
+		pr_err("parse madt failed\n");
```
Comment (gvancuts) [inline devicemodel/core/main.c:850]:
```suggestion
		pr_err("Failed to parse the MADT table.\n");
```

File: devicemodel/hw/platform/acpi/acpi.c [hunk 4]
```diff
@@ -343,7 +338,11 @@ basl_fwrite_madt(FILE *fp, struct vmctx *ctx)
 			return -1;
 		}
 
-		lapic_id = lapicid_from_pcpuid(&plat_info, pcpu_id);
+		lapic_id = lapicid_from_pcpuid(pcpu_id);
+		if (lapic_id == -1) {
+			pr_err("Get lapic id fail.\n");
```
Comment (gvancuts) [inline devicemodel/hw/platform/acpi/acpi.c:343]:
```suggestion
			pr_err("Failed to retrieve the local APIC IDs for pCPU %u\n", pcpu_id);
```

File: devicemodel/hw/platform/acpi/acpi_parser.c [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+	if (madt == NULL) {
+		pr_err("Failed to malloc %d bytes for MADT!\n", file_state.st_size);
+		close(fd);
+		return -1;
+	}
+	size = read(fd, madt, file_state.st_size);
+	if (size == file_state.st_size) {
+		ret = local_parse_madt(madt);
+	} else {
+		pr_err("Fail to read Service VM madt info!");
```
Comment (gvancuts) [inline devicemodel/hw/platform/acpi/acpi_parser.c:143]:
```suggestion
		pr_err("Failed to read Service VM MADT table.");
```

File: devicemodel/hw/platform/acpi/acpi_parser.c [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+
+	if (fstat(fd, &file_state) == -1) {
+		pr_err("Fail to get apic file state!\n");
+		close(fd);
+		return -1;
+	}
+
+	madt = (struct acpi_table_madt *)malloc(file_state.st_size);
+	if (madt == NULL) {
+		pr_err("Failed to malloc %d bytes for MADT!\n", file_state.st_size);
```
Comment (gvancuts) [inline devicemodel/hw/platform/acpi/acpi_parser.c:135]:
```suggestion
		pr_err("Failed to malloc %d bytes to store the MADT table.\n", file_state.st_size);
```

File: devicemodel/hw/platform/acpi/acpi_parser.c [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+	struct stat file_state;
+
+	int fd = open("/sys/firmware/acpi/tables/APIC", O_RDONLY);
+	if (fd < 0) {
+		pr_err("Fail to open Service VM APIC file!\n");
+		return -1;
+	}
+
+	if (fstat(fd, &file_state) == -1) {
+		pr_err("Fail to get apic file state!\n");
```
Comment (gvancuts) [inline devicemodel/hw/platform/acpi/acpi_parser.c:128]:
```suggestion
		pr_err("Failed to get APIC file state.\n");
```

File: devicemodel/hw/platform/acpi/acpi_parser.c [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+int parse_madt(void)
+{
+	int ret = 0U;
+	ssize_t size;
+	struct acpi_table_madt *madt;
+	struct stat file_state;
+
+	int fd = open("/sys/firmware/acpi/tables/APIC", O_RDONLY);
+	if (fd < 0) {
+		pr_err("Fail to open Service VM APIC file!\n");
```
Comment (gvancuts) [inline devicemodel/hw/platform/acpi/acpi_parser.c:123]:
```suggestion
		pr_err("Failed to open the Service VM APIC file.\n");
```

File: devicemodel/hw/platform/acpi/acpi_parser.c [hunk 9]
```diff
@@ -0,0 +1,150 @@
+/*-
+ * Copyright (c) 2011 NetApp, Inc.
```
Comment (gvancuts) [inline devicemodel/hw/platform/acpi/acpi_parser.c:2]: Please see and answer my question from the mailing list.

**Review by gvancuts — CHANGES_REQUESTED** [review]
A number of questions and comments had not been addressed before this was sent as a PR, why is that?

File: devicemodel/core/main.c (hunk 1 above)
Comment (YuanyuanZhao609) [reply devicemodel/core/main.c:169]: OK，will change it next version.

File: devicemodel/core/main.c (hunk 2 above)
Comment (YuanyuanZhao609) [reply devicemodel/core/main.c:170]: OK，will change it next version.

File: devicemodel/core/main.c (hunk 3 above)
Comment (YuanyuanZhao609) [reply devicemodel/core/main.c:850]: OK，will change it next version.

File: devicemodel/hw/platform/acpi/acpi.c (hunk 4 above)
Comment (YuanyuanZhao609) [reply devicemodel/hw/platform/acpi/acpi.c:343]: OK，will change it next version.

File: devicemodel/hw/platform/acpi/acpi_parser.c (hunk 5 above)
Comment (YuanyuanZhao609) [reply devicemodel/hw/platform/acpi/acpi_parser.c:143]: OK，will change it next version.

File: devicemodel/hw/platform/acpi/acpi_parser.c (hunk 6 above)
Comment (YuanyuanZhao609) [reply devicemodel/hw/platform/acpi/acpi_parser.c:135]: OK，will change it next version.

File: devicemodel/hw/platform/acpi/acpi_parser.c (hunk 7 above)
Comment (YuanyuanZhao609) [reply devicemodel/hw/platform/acpi/acpi_parser.c:128]: OK，will change it next version.

File: devicemodel/hw/platform/acpi/acpi_parser.c (hunk 8 above)
Comment (YuanyuanZhao609) [reply devicemodel/hw/platform/acpi/acpi_parser.c:123]: OK，will change it next version.

File: devicemodel/hw/platform/acpi/acpi_parser.c (hunk 9 above)
Comment (YuanyuanZhao609) [reply devicemodel/hw/platform/acpi/acpi_parser.c:2]: OK

Comment (YuanyuanZhao609) [reply devicemodel/hw/platform/acpi/acpi_parser.c:2]: I will change it to 'Copyright (C) 2021 Intel Corporation. '

File: devicemodel/hw/platform/acpi/acpi_parser.c [hunk 10]
```diff
@@ -0,0 +1,129 @@
+/*-
+ * Copyright (C) 2021 Intel Corporation. All rights reserved.
```
Comment (gvancuts) [inline devicemodel/hw/platform/acpi/acpi_parser.c:2]:
```suggestion
 * Copyright (C) 2022 Intel Corporation.
```

Comment (gvancuts) [reply devicemodel/hw/platform/acpi/acpi_parser.c:2]: Was that file originally from FreeBSD? In other words, is it appropriate to completely remove the original Copyright notice that was in here?

Comment (YuanyuanZhao609) [reply devicemodel/hw/platform/acpi/acpi_parser.c:2]: In the beginning, the whole acpi parser file is imported, but in the end only one function(local_parse_madt) is kept.
Should the copyright of FreeBSD be announced?

Comment (dbkinder) [reply devicemodel/hw/platform/acpi/acpi_parser.c:2]: Original copyright must remain but you can add a second copyright line.

Comment (YuanyuanZhao609) [reply devicemodel/hw/platform/acpi/acpi_parser.c:2]: Updated

File: devicemodel/hw/platform/acpi/acpi.c [hunk 11]
```diff
@@ -343,7 +338,11 @@ basl_fwrite_madt(FILE *fp, struct vmctx *ctx)
 			return -1;
 		}
 
-		lapic_id = lapicid_from_pcpuid(&plat_info, pcpu_id);
+		lapic_id = lapicid_from_pcpuid(pcpu_id);
+		if (lapic_id == -1) {
+			pr_err("Failed to retrieve the local APIC IDs for pCPU %u\n", pcpu_id);
```
[the remark this answers is not in the document]
Comment (YuanyuanZhao609) [reply devicemodel/hw/platform/acpi/acpi.c:343]: OK

File: devicemodel/hw/platform/acpi/acpi_parser.c [hunk 12]
```diff
@@ -0,0 +1,151 @@
+/*-
+ * SPDX-License-Identifier: BSD-2-Clause-FreeBSD
```
[the remark this answers is not in the document]
Comment (YuanyuanZhao609) [reply devicemodel/hw/platform/acpi/acpi_parser.c:2]: OK.

This statement is copied from the original document. For non-professionals in this field, the original statement is copied in most cases.

[the remark this answers is not in the document]
Comment (gvancuts) [reply devicemodel/hw/platform/acpi/acpi_parser.c:2]: I'm not sure I'm understanding your point.

It's fine to copy the text below as is. I do believe you have added the SPDX identifier yourself though, haven't you? That identifier seems wrong based on the text which is the point I'm raising here.

**Review by gvancuts — APPROVED** [review]

**Comment by wenlingz** [comment]
try verify EHL

**Comment by wenlingz** [comment]
start to run premerge test

**Comment by wenlingz** [comment]
@ywan170 please help approve this PR, thanks

**Review by ywan170 — APPROVED** [review]

**Comment by wenlingz** [comment]
@Kunhui-Li please rebase code, thanks

**Pull request merged**